Looks like the linter is unhappy (because use of unkeyed literals); not sure if you're using GoLand, but that one can "fill in all keys" for you;

```
libnetwork/drivers/bridge/setup_ip_tables_linux_test.go:418:23: composites: github.com/docker/docker/libnetwork/iptables.Rule struct literal uses unkeyed fields (govet)
				{tc.wantIPv4Masq, iptables.Rule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
				                  ^
libnetwork/drivers/bridge/setup_ip_tables_linux_test.go:419:23: composites: github.com/docker/docker/libnetwork/iptables.Rule struct literal uses unkeyed fields (govet)
				{tc.wantIPv4Snat, iptables.Rule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv4.String()}}},
				                  ^
libnetwork/drivers/bridge/setup_ip_tables_linux_test.go:420:23: composites: github.com/docker/docker/libnetwork/iptables.Rule struct literal uses unkeyed fields (govet)
				{tc.wantIPv6Masq, iptables.Rule{iptables.IPv6, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
				                  ^
libnetwork/drivers/bridge/setup_ip_tables_linux_test.go:421:23: composites: github.com/docker/docker/libnetwork/iptables.Rule struct literal uses unkeyed fields (govet)
				{tc.wantIPv6Snat, iptables.Rule{iptables.IPv6, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv6.String()}}},
				                  ^
```

--- a/libnetwork/drivers/bridge/port_mapping_linux.go
+++ b/libnetwork/drivers/bridge/port_mapping_linux.go
@@ -811,19 +811,18 @@ func setPerPortNAT(b portBinding, ipv iptables.IPVersion, proxyPath string, brid
 	if ipv == iptables.IPv6 {
 		args = append(args, "!", "-s", "fe80::/10")
 	}
-	rule := iptRule{ipv: ipv, table: iptables.Nat, chain: DockerChain, args: args}
+	rule := iptables.Rule{IPVer: ipv, Table: iptables.Nat, Chain: DockerChain, Args: args}
 	if err := appendOrDelChainRule(rule, "DNAT", enable); err != nil {
 		return err
 	}
 
-	args = []string{
+	rule = iptables.Rule{IPVer: ipv, Table: iptables.Nat, Chain: "POSTROUTING", Args: []string{
 		"-p", b.Proto.String(),
 		"-s", b.IP.String(),
 		"-d", b.IP.String(),
 		"--dport", strconv.Itoa(int(b.Port)),
 		"-j", "MASQUERADE",
-	}
-	rule = iptRule{ipv: ipv, table: iptables.Nat, chain: "POSTROUTING", args: args}
+	}}
 	if err := appendOrDelChainRule(rule, "MASQUERADE", hairpinMode && enable); err != nil {
 		return err
 	}
@@ -836,15 +835,14 @@ func setPerPortForwarding(b portBinding, ipv iptables.IPVersion, bridgeName stri
 	// chain (a per-network DROP rule, which must come after these per-port
 	// per-container ACCEPT rules, is appended to the chain when the network
 	// is created).
-	args := []string{
+	rule := iptables.Rule{IPVer: ipv, Table: iptables.Filter, Chain: DockerChain, Args: []string{
 		"!", "-i", bridgeName,
 		"-o", bridgeName,
 		"-p", b.Proto.String(),
 		"-d", b.IP.String(),
 		"--dport", strconv.Itoa(int(b.Port)),
 		"-j", "ACCEPT",
-	}
-	rule := iptRule{ipv: ipv, table: iptables.Filter, chain: DockerChain, args: args}
+	}}
 	if err := programChainRule(rule, "OPEN PORT", enable); err != nil {
 		return err
 	}
@@ -857,13 +855,12 @@ func setPerPortForwarding(b portBinding, ipv iptables.IPVersion, bridgeName stri
 		// to fill the checksum.
 		//
 		// https://github.com/torvalds/linux/commit/c80fafbbb59ef9924962f83aac85531039395b18
-		args = []string{
+		rule := iptables.Rule{IPVer: ipv, Table: iptables.Mangle, Chain: "POSTROUTING", Args: []string{
 			"-p", b.Proto.String(),
 			"--sport", strconv.Itoa(int(b.Port)),
 			"-j", "CHECKSUM",
 			"--checksum-fill",
-		}
-		rule := iptRule{ipv: ipv, table: iptables.Mangle, chain: "POSTROUTING", args: args}
+		}}
 		if err := appendOrDelChainRule(rule, "SCTP CHECKSUM", enable); err != nil {
 			return err
 		}

--- a/libnetwork/drivers/bridge/port_mapping_linux_test.go
+++ b/libnetwork/drivers/bridge/port_mapping_linux_test.go
@@ -903,7 +903,7 @@ func TestAddPortMappings(t *testing.T) {
 				// Check the MASQUERADE rule.
 				masqRule := fmt.Sprintf("-s %s -d %s -p %s -m %s --dport %d -j MASQUERADE",
 					addrM, addrM, expPB.Proto, expPB.Proto, expPB.Port)
-				ir := iptRule{ipv: ipv, table: iptables.Nat, chain: "POSTROUTING", args: strings.Split(masqRule, " ")}
+				ir := iptables.Rule{IPVer: ipv, Table: iptables.Nat, Chain: "POSTROUTING", Args: strings.Split(masqRule, " ")}
 				if disableNAT || tc.proxyPath != "" {
 					assert.Check(t, !ir.Exists(), fmt.Sprintf("unexpected rule %s", ir))
 				} else {
@@ -921,7 +921,7 @@ func TestAddPortMappings(t *testing.T) {
 				}
 				dnatRule += fmt.Sprintf("-d %s -p %s -m %s --dport %d -j DNAT --to-destination %s:%d",
 					addrH, expPB.Proto, expPB.Proto, expPB.HostPort, addrD, expPB.Port)
-				ir = iptRule{ipv: ipv, table: iptables.Nat, chain: "DOCKER", args: strings.Split(dnatRule, " ")}
+				ir = iptables.Rule{IPVer: ipv, Table: iptables.Nat, Chain: "DOCKER", Args: strings.Split(dnatRule, " ")}
 				if disableNAT {
 					assert.Check(t, !ir.Exists(), fmt.Sprintf("unexpected rule %s", ir))
 				} else {
@@ -931,7 +931,7 @@ func TestAddPortMappings(t *testing.T) {
 				// Check that the container's port is open.
 				filterRule := fmt.Sprintf("-d %s ! -i dummybridge -o dummybridge -p %s -m %s --dport %d -j ACCEPT",
 					addrM, expPB.Proto, expPB.Proto, expPB.Port)
-				ir = iptRule{ipv: ipv, table: iptables.Filter, chain: "DOCKER", args: strings.Split(filterRule, " ")}
+				ir = iptables.Rule{IPVer: ipv, Table: iptables.Filter, Chain: "DOCKER", Args: strings.Split(filterRule, " ")}
 				assert.Check(t, ir.Exists(), fmt.Sprintf("expected rule %s", ir))
 			}
 

--- a/libnetwork/drivers/bridge/setup_ip_tables_linux.go
+++ b/libnetwork/drivers/bridge/setup_ip_tables_linux.go
@@ -6,7 +6,6 @@ import (
 	"fmt"
 	"net"
 	"os"
-	"strings"
 
 	"github.com/containerd/log"
 	"github.com/docker/docker/errdefs"
@@ -253,7 +252,7 @@ func setICMP(ipv iptables.IPVersion, bridgeName string, enable bool) error {
 	if ipv == iptables.IPv6 {
 		icmpProto = "icmpv6"
 	}
-	icmpRule := iptRule{ipv: ipv, table: iptables.Filter, chain: DockerChain, args: []string{
+	icmpRule := iptables.Rule{IPVer: ipv, Table: iptables.Filter, Chain: DockerChain, Args: []string{
 		"-o", bridgeName,
 		"-p", icmpProto,
 		"-j", "ACCEPT",
@@ -262,7 +261,7 @@ func setICMP(ipv iptables.IPVersion, bridgeName string, enable bool) error {
 }
 
 func addNATJumpRules(ipVer iptables.IPVersion, hairpinMode, enable bool) error {
-	preroute := iptRule{ipv: ipVer, table: iptables.Nat, chain: "PREROUTING", args: []string{
+	preroute := iptables.Rule{IPVer: ipVer, Table: iptables.Nat, Chain: "PREROUTING", Args: []string{
 		"-m", "addrtype",
 		"--dst-type", "LOCAL",
 		"-j", DockerChain,
@@ -277,13 +276,13 @@ func addNATJumpRules(ipVer iptables.IPVersion, hairpinMode, enable bool) error {
 		}
 	}
 
-	output := iptRule{ipv: ipVer, table: iptables.Nat, chain: "OUTPUT", args: []string{
+	output := iptables.Rule{IPVer: ipVer, Table: iptables.Nat, Chain: "OUTPUT", Args: []string{
 		"-m", "addrtype",
 		"--dst-type", "LOCAL",
 		"-j", DockerChain,
 	}}
 	if !hairpinMode {
-		output.args = append(output.args, "!", "--dst", loopbackAddress(ipVer))
+		output.Args = append(output.Args, "!", "--dst", loopbackAddress(ipVer))
 	}
 	if enable {
 		if err := output.Append(); err != nil {
@@ -367,7 +366,7 @@ func (n *bridgeNetwork) setDefaultForwardRule(
 		action = "ACCEPT"
 	}
 
-	rule := iptRule{ipv: ipVersion, table: iptables.Filter, chain: DockerChain, args: []string{
+	rule := iptables.Rule{IPVer: ipVersion, Table: iptables.Filter, Chain: DockerChain, Args: []string{
 		"!", "-i", bridgeName,
 		"-o", bridgeName,
 		"-j", action,
@@ -382,60 +381,6 @@ func (n *bridgeNetwork) setDefaultForwardRule(
 		return appendOrDelChainRule(rule, "DEFAULT FWD", false)
 	})
 	return nil
-}
-
-type iptRule struct {
-	ipv   iptables.IPVersion
-	table iptables.Table
-	chain string
-	args  []string
-}
-
-// Exists returns true if the rule exists in the kernel.
-func (r iptRule) Exists() bool {
-	return iptables.GetIptable(r.ipv).Exists(r.table, r.chain, r.args...)
-}
-
-func (r iptRule) cmdArgs(op iptables.Action) []string {
-	return append([]string{"-t", string(r.table), string(op), r.chain}, r.args...)
-}
-
-func (r iptRule) exec(op iptables.Action) error {
-	return iptables.GetIptable(r.ipv).RawCombinedOutput(r.cmdArgs(op)...)
-}
-
-// Append appends the rule to the end of the chain. If the rule already exists anywhere in the
-// chain, this is a no-op.
-func (r iptRule) Append() error {
-	if r.Exists() {
-		return nil
-	}
-	return r.exec(iptables.Append)
-}
-
-// Insert inserts the rule at the head of the chain. If the rule already exists anywhere in the
-// chain, this is a no-op.
-func (r iptRule) Insert() error {
-	if r.Exists() {
-		return nil
-	}
-	return r.exec(iptables.Insert)
-}
-
-// Delete deletes the rule from the kernel. If the rule does not exist, this is a no-op.
-func (r iptRule) Delete() error {
-	if !r.Exists() {
-		return nil
-	}
-	return r.exec(iptables.Delete)
-}
-
-func (r iptRule) String() string {
-	cmd := append([]string{"iptables"}, r.cmdArgs("-A")...)
-	if r.ipv == iptables.IPv6 {
-		cmd[0] = "ip6tables"
-	}
-	return strings.Join(cmd, " ")
 }
 
 func setupNonInternalNetworkRules(ipVer iptables.IPVersion, config *networkConfiguration, addr *net.IPNet, hairpin, enable bool) error {
@@ -457,8 +402,8 @@ func setupNonInternalNetworkRules(ipVer iptables.IPVersion, config *networkConfi
 		natArgs = []string{"-s", addr.String(), "!", "-o", config.BridgeName, "-j", "MASQUERADE"}
 		hpNatArgs = []string{"-m", "addrtype", "--src-type", "LOCAL", "-o", config.BridgeName, "-j", "MASQUERADE"}
 	}
-	natRule := iptRule{ipv: ipVer, table: iptables.Nat, chain: "POSTROUTING", args: natArgs}
-	hpNatRule := iptRule{ipv: ipVer, table: iptables.Nat, chain: "POSTROUTING", args: hpNatArgs}
+	natRule := iptables.Rule{IPVer: ipVer, Table: iptables.Nat, Chain: "POSTROUTING", Args: natArgs}
+	hpNatRule := iptables.Rule{IPVer: ipVer, Table: iptables.Nat, Chain: "POSTROUTING", Args: hpNatArgs}
 
 	// Set NAT.
 	if nat && config.EnableIPMasquerade {
@@ -467,7 +412,7 @@ func setupNonInternalNetworkRules(ipVer iptables.IPVersion, config *networkConfi
 		}
 	}
 	if !nat || (config.EnableIPMasquerade && !hairpin) {
-		skipDNAT := iptRule{ipv: ipVer, table: iptables.Nat, chain: DockerChain, args: []string{
+		skipDNAT := iptables.Rule{IPVer: ipVer, Table: iptables.Nat, Chain: DockerChain, Args: []string{
 			"-i", config.BridgeName,
 			"-j", "RETURN",
 		}}
@@ -498,7 +443,7 @@ func setupNonInternalNetworkRules(ipVer iptables.IPVersion, config *networkConfi
 	// to ACCEPT packets that weren't ICC - an extra rule was needed to enable
 	// ICC if needed. Those rules are now combined. So, outRuleNoICC is only
 	// needed for ICC=false, along with the DROP rule for ICC added by setIcc.
-	outRuleNoICC := iptRule{ipv: ipVer, table: iptables.Filter, chain: "FORWARD", args: []string{
+	outRuleNoICC := iptables.Rule{IPVer: ipVer, Table: iptables.Filter, Chain: "FORWARD", Args: []string{
 		"-i", config.BridgeName,
 		"!", "-o", config.BridgeName,
 		"-j", "ACCEPT",
@@ -510,7 +455,7 @@ func setupNonInternalNetworkRules(ipVer iptables.IPVersion, config *networkConfi
 			return err
 		}
 		// Accept outgoing traffic to anywhere, including other containers on this bridge.
-		outRuleICC := iptRule{ipv: ipVer, table: iptables.Filter, chain: "FORWARD", args: []string{
+		outRuleICC := iptables.Rule{IPVer: ipVer, Table: iptables.Filter, Chain: "FORWARD", Args: []string{
 			"-i", config.BridgeName,
 			"-j", "ACCEPT",
 		}}
@@ -528,7 +473,7 @@ func setupNonInternalNetworkRules(ipVer iptables.IPVersion, config *networkConfi
 	return nil
 }
 
-func programChainRule(rule iptRule, ruleDescr string, insert bool) error {
+func programChainRule(rule iptables.Rule, ruleDescr string, insert bool) error {
 	operation := "disable"
 	fn := rule.Delete
 	if insert {
@@ -541,7 +486,7 @@ func programChainRule(rule iptRule, ruleDescr string, insert bool) error {
 	return nil
 }
 
-func appendOrDelChainRule(rule iptRule, ruleDescr string, append bool) error {
+func appendOrDelChainRule(rule iptables.Rule, ruleDescr string, append bool) error {
 	operation := "disable"
 	fn := rule.Delete
 	if append {
@@ -556,8 +501,8 @@ func appendOrDelChainRule(rule iptRule, ruleDescr string, append bool) error {
 
 func setIcc(version iptables.IPVersion, bridgeIface string, iccEnable, internal, insert bool) error {
 	args := []string{"-i", bridgeIface, "-o", bridgeIface, "-j"}
-	acceptRule := iptRule{ipv: version, table: iptables.Filter, chain: "FORWARD", args: append(args, "ACCEPT")}
-	dropRule := iptRule{ipv: version, table: iptables.Filter, chain: "FORWARD", args: append(args, "DROP")}
+	acceptRule := iptables.Rule{IPVer: version, Table: iptables.Filter, Chain: "FORWARD", Args: append(args, "ACCEPT")}
+	dropRule := iptables.Rule{IPVer: version, Table: iptables.Filter, Chain: "FORWARD", Args: append(args, "DROP")}
 
 	// The accept rule is no longer required for a bridge with external connectivity, because
 	// ICC traffic is allowed by the outgoing-packets rule created by setupIptablesInternal.
@@ -675,7 +620,7 @@ func removeIPChains(version iptables.IPVersion) {
 
 func setupInternalNetworkRules(bridgeIface string, addr *net.IPNet, icc, insert bool) error {
 	var version iptables.IPVersion
-	var inDropRule, outDropRule iptRule
+	var inDropRule, outDropRule iptables.Rule
 
 	// Either add or remove the interface from the firewalld zone, if firewalld is running.
 	if insert {
@@ -690,31 +635,31 @@ func setupInternalNetworkRules(bridgeIface string, addr *net.IPNet, icc, insert 
 
 	if addr.IP.To4() != nil {
 		version = iptables.IPv4
-		inDropRule = iptRule{
-			ipv:   version,
-			table: iptables.Filter,
-			chain: IsolationChain1,
-			args:  []string{"-i", bridgeIface, "!", "-d", addr.String(), "-j", "DROP"},
+		inDropRule = iptables.Rule{
+			IPVer: version,
+			Table: iptables.Filter,
+			Chain: IsolationChain1,
+			Args:  []string{"-i", bridgeIface, "!", "-d", addr.String(), "-j", "DROP"},
 		}
-		outDropRule = iptRule{
-			ipv:   version,
-			table: iptables.Filter,
-			chain: IsolationChain1,
-			args:  []string{"-o", bridgeIface, "!", "-s", addr.String(), "-j", "DROP"},
+		outDropRule = iptables.Rule{
+			IPVer: version,
+			Table: iptables.Filter,
+			Chain: IsolationChain1,
+			Args:  []string{"-o", bridgeIface, "!", "-s", addr.String(), "-j", "DROP"},
 		}
 	} else {
 		version = iptables.IPv6
-		inDropRule = iptRule{
-			ipv:   version,
-			table: iptables.Filter,
-			chain: IsolationChain1,
-			args:  []string{"-i", bridgeIface, "!", "-o", bridgeIface, "!", "-d", addr.String(), "-j", "DROP"},
+		inDropRule = iptables.Rule{
+			IPVer: version,
+			Table: iptables.Filter,
+			Chain: IsolationChain1,
+			Args:  []string{"-i", bridgeIface, "!", "-o", bridgeIface, "!", "-d", addr.String(), "-j", "DROP"},
 		}
-		outDropRule = iptRule{
-			ipv:   version,
-			table: iptables.Filter,
-			chain: IsolationChain1,
-			args:  []string{"!", "-i", bridgeIface, "-o", bridgeIface, "!", "-s", addr.String(), "-j", "DROP"},
+		outDropRule = iptables.Rule{
+			IPVer: version,
+			Table: iptables.Filter,
+			Chain: IsolationChain1,
+			Args:  []string{"!", "-i", bridgeIface, "-o", bridgeIface, "!", "-s", addr.String(), "-j", "DROP"},
 		}
 	}
 
@@ -831,11 +776,11 @@ func insertMirroredWSL2Rule(config configuration) bool {
 	return stat.Mode().IsRegular() && (stat.Mode().Perm()&0o111) != 0
 }
 
-func mirroredWSL2Rule() iptRule {
-	return iptRule{
-		ipv:   iptables.IPv4,
-		table: iptables.Nat,
-		chain: DockerChain,
-		args:  []string{"-i", "loopback0", "-d", "127.0.0.0/8", "-j", "RETURN"},
+func mirroredWSL2Rule() iptables.Rule {
+	return iptables.Rule{
+		IPVer: iptables.IPv4,
+		Table: iptables.Nat,
+		Chain: DockerChain,
+		Args:  []string{"-i", "loopback0", "-d", "127.0.0.0/8", "-j", "RETURN"},
 	}
 }

--- a/libnetwork/drivers/bridge/setup_ip_tables_linux_test.go
+++ b/libnetwork/drivers/bridge/setup_ip_tables_linux_test.go
@@ -52,15 +52,15 @@ func TestProgramIPTable(t *testing.T) {
 
 	// Store various iptables chain rules we care for.
 	rules := []struct {
-		rule  iptRule
+		rule  iptables.Rule
 		descr string
 	}{
-		{iptRule{ipv: iptables.IPv4, table: iptables.Filter, chain: "FORWARD", args: []string{"-d", "127.1.2.3", "-i", "lo", "-o", "lo", "-j", "DROP"}}, "Test Loopback"},
-		{iptRule{ipv: iptables.IPv4, table: iptables.Nat, chain: "POSTROUTING", args: []string{"-s", iptablesTestBridgeIP, "!", "-o", DefaultBridgeName, "-j", "MASQUERADE"}}, "NAT Test"},
-		{iptRule{ipv: iptables.IPv4, table: iptables.Filter, chain: "FORWARD", args: []string{"-o", DefaultBridgeName, "-m", "conntrack", "--ctstate", "RELATED,ESTABLISHED", "-j", "ACCEPT"}}, "Test ACCEPT INCOMING"},
-		{iptRule{ipv: iptables.IPv4, table: iptables.Filter, chain: "FORWARD", args: []string{"-i", DefaultBridgeName, "!", "-o", DefaultBridgeName, "-j", "ACCEPT"}}, "Test ACCEPT NON_ICC OUTGOING"},
-		{iptRule{ipv: iptables.IPv4, table: iptables.Filter, chain: "FORWARD", args: []string{"-i", DefaultBridgeName, "-o", DefaultBridgeName, "-j", "ACCEPT"}}, "Test enable ICC"},
-		{iptRule{ipv: iptables.IPv4, table: iptables.Filter, chain: "FORWARD", args: []string{"-i", DefaultBridgeName, "-o", DefaultBridgeName, "-j", "DROP"}}, "Test disable ICC"},
+		{iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Filter, Chain: "FORWARD", Args: []string{"-d", "127.1.2.3", "-i", "lo", "-o", "lo", "-j", "DROP"}}, "Test Loopback"},
+		{iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Nat, Chain: "POSTROUTING", Args: []string{"-s", iptablesTestBridgeIP, "!", "-o", DefaultBridgeName, "-j", "MASQUERADE"}}, "NAT Test"},
+		{iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Filter, Chain: "FORWARD", Args: []string{"-o", DefaultBridgeName, "-m", "conntrack", "--ctstate", "RELATED,ESTABLISHED", "-j", "ACCEPT"}}, "Test ACCEPT INCOMING"},
+		{iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Filter, Chain: "FORWARD", Args: []string{"-i", DefaultBridgeName, "!", "-o", DefaultBridgeName, "-j", "ACCEPT"}}, "Test ACCEPT NON_ICC OUTGOING"},
+		{iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Filter, Chain: "FORWARD", Args: []string{"-i", DefaultBridgeName, "-o", DefaultBridgeName, "-j", "ACCEPT"}}, "Test enable ICC"},
+		{iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Filter, Chain: "FORWARD", Args: []string{"-i", DefaultBridgeName, "-o", DefaultBridgeName, "-j", "DROP"}}, "Test disable ICC"},
 	}
 
 	// Assert the chain rules' insertion and removal.
@@ -133,7 +133,7 @@ func createTestBridge(config *networkConfiguration, br *bridgeInterface, t *test
 }
 
 // Assert base function which pushes iptables chain rules on insertion and removal.
-func assertIPTableChainProgramming(rule iptRule, descr string, t *testing.T) {
+func assertIPTableChainProgramming(rule iptables.Rule, descr string, t *testing.T) {
 	// Add
 	if err := programChainRule(rule, descr, true); err != nil {
 		t.Fatalf("Failed to program iptable rule %s: %s", descr, err.Error())
@@ -411,14 +411,14 @@ func TestOutgoingNATRules(t *testing.T) {
 			}
 			for _, rc := range []struct {
 				want bool
-				rule iptRule
+				rule iptables.Rule
 			}{
 				// Rule order doesn't matter: At most one of the following IPv4 rules will exist, and the
 				// same goes for the IPv6 rules.
-				{tc.wantIPv4Masq, iptRule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
-				{tc.wantIPv4Snat, iptRule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv4.String()}}},
-				{tc.wantIPv6Masq, iptRule{iptables.IPv6, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
-				{tc.wantIPv6Snat, iptRule{iptables.IPv6, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv6.String()}}},
+				{tc.wantIPv4Masq, iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Nat, Chain: "POSTROUTING", Args: []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
+				{tc.wantIPv4Snat, iptables.Rule{IPVer: iptables.IPv4, Table: iptables.Nat, Chain: "POSTROUTING", Args: []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv4.String()}}},
+				{tc.wantIPv6Masq, iptables.Rule{IPVer: iptables.IPv6, Table: iptables.Nat, Chain: "POSTROUTING", Args: []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
+				{tc.wantIPv6Snat, iptables.Rule{IPVer: iptables.IPv6, Table: iptables.Nat, Chain: "POSTROUTING", Args: []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv6.String()}}},
 			} {
 				assert.Equal(t, rc.rule.Exists(), rc.want)
 			}

--- a/libnetwork/iptables/iptables.go
+++ b/libnetwork/iptables/iptables.go
@@ -424,3 +424,57 @@ func (iptable IPTable) EnsureJumpRule(fromChain, toChain string, rule ...string)
 	}
 	return nil
 }
+
+type Rule struct {
+	IPVer IPVersion
+	Table Table
+	Chain string
+	Args  []string
+}
+
+// Exists returns true if the rule exists in the kernel.
+func (r Rule) Exists() bool {
+	return GetIptable(r.IPVer).Exists(r.Table, r.Chain, r.Args...)
+}
+
+func (r Rule) cmdArgs(op Action) []string {
+	return append([]string{"-t", string(r.Table), string(op), r.Chain}, r.Args...)
+}
+
+func (r Rule) exec(op Action) error {
+	return GetIptable(r.IPVer).RawCombinedOutput(r.cmdArgs(op)...)
+}
+
+// Append appends the rule to the end of the chain. If the rule already exists anywhere in the
+// chain, this is a no-op.
+func (r Rule) Append() error {
+	if r.Exists() {
+		return nil
+	}
+	return r.exec(Append)
+}
+
+// Insert inserts the rule at the head of the chain. If the rule already exists anywhere in the
+// chain, this is a no-op.
+func (r Rule) Insert() error {
+	if r.Exists() {
+		return nil
+	}
+	return r.exec(Insert)
+}
+
+// Delete deletes the rule from the kernel. If the rule does not exist, this is a no-op.
+func (r Rule) Delete() error {
+	if !r.Exists() {
+		return nil
+	}
+	return r.exec(Delete)
+}
+
+func (r Rule) String() string {
+	cmd := append([]string{"iptables"}, r.cmdArgs("-A")...)
+	if r.IPVer == IPv6 {
+		cmd[0] = "ip6tables"
+	}
+	return strings.Join(cmd, " ")
+}

--- a/libnetwork/iptables/iptables_test.go
+++ b/libnetwork/iptables/iptables_test.go
@@ -9,7 +9,9 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/docker/docker/internal/testutils/netnsutils"
 	"golang.org/x/sync/errgroup"
+	"gotest.tools/v3/assert"
 )
 
 const (
@@ -252,4 +254,54 @@ func TestExistsRaw(t *testing.T) {
 			t.Fatalf("Invalid detection. i=%d", i)
 		}
 	}
+}
+
+func TestRule(t *testing.T) {
+	defer netnsutils.SetupTestOSContext(t)()
+
+	assert.NilError(t, exec.Command("iptables", "-N", "TESTCHAIN").Run())
+
+	rule := Rule{IPVer: IPv4, Table: Filter, Chain: "TESTCHAIN", Args: []string{"-j", "RETURN"}}
+	assert.NilError(t, rule.Insert())
+	assert.Equal(t, rule.Exists(), true)
+	assert.Equal(t, mustDumpChain(t, Filter, "TESTCHAIN"), `-N TESTCHAIN
+-A TESTCHAIN -j RETURN
+`)
+
+	// Test that Insert is idempotent
+	assert.NilError(t, rule.Insert())
+	assert.Equal(t, mustDumpChain(t, Filter, "TESTCHAIN"), `-N TESTCHAIN
+-A TESTCHAIN -j RETURN
+`)
+
+	rule = Rule{IPVer: IPv4, Table: Filter, Chain: "TESTCHAIN", Args: []string{"-j", "ACCEPT"}}
+	assert.NilError(t, rule.Append())
+	assert.Equal(t, rule.Exists(), true)
+	assert.Equal(t, mustDumpChain(t, Filter, "TESTCHAIN"), `-N TESTCHAIN
+-A TESTCHAIN -j RETURN
+-A TESTCHAIN -j ACCEPT
+`)
+
+	// Test that Append is idempotent
+	assert.NilError(t, rule.Append())
+	assert.Equal(t, mustDumpChain(t, Filter, "TESTCHAIN"), `-N TESTCHAIN
+-A TESTCHAIN -j RETURN
+-A TESTCHAIN -j ACCEPT
+`)
+
+	assert.NilError(t, rule.Delete())
+	assert.Equal(t, rule.Exists(), false)
+	assert.Equal(t, mustDumpChain(t, Filter, "TESTCHAIN"), `-N TESTCHAIN
+-A TESTCHAIN -j RETURN
+`)
+
+	// Test that Delete is idempotent
+	assert.NilError(t, rule.Delete())
+}
+
+func mustDumpChain(t *testing.T, table Table, chain string) string {
+	t.Helper()
+	out, err := exec.Command("iptables", "-t", string(table), "-S", chain).CombinedOutput()
+	assert.NilError(t, err, "output:\n%s", out)
+	return string(out)
 }

